### PR TITLE
[SPARK-19554][UI,YARN] Allow SHS URL to be used for tracking in YARN RM.

### PR DESCRIPTION
**Pull request by vanzin**
Allow an application to use the History Server URL as the tracking
URL in the YARN RM, so there's still a link to the web UI somewhere
in YARN even if the driver's UI is disabled. This is useful, for
example, if an admin wants to disable the driver UI by default for
applications, since it's harder to secure it (since it involves non
trivial ssl certificate and auth management that admins may not want
to expose to user apps).

This needs to be opt-in, because of the way the YARN proxy works, so
a new configuration was added to enable the option.

The YARN RM will proxy requests to live AMs instead of redirecting
the client, so pages in the SHS UI will not render correctly since
they'll reference invalid paths in the RM UI. The proxy base support
in the SHS cannot be used since that would prevent direct access to
the SHS.

So, to solve this problem, for the feature to work end-to-end, a new
YARN-specific filter was added that detects whether the requests come
from the proxy and redirects the client appropriatly. The SHS admin has
to add this filter manually if they want the feature to work.

Tested with new unit test, and by running with the documented configuration
set in a test cluster. Also verified the driver UI is used when it's
enabled.

### CODE REVIEW

**Comment by vanzin** [comment]
@tgravescs pinging you since this is all YARN-side (doesn't really touch the UI).

**Comment by vanzin** [comment]
trying to get some eyes on this: @squito

**Comment by tgravescs** [comment]
On vacation back next Monday and will review.

File: docs/running-on-yarn.md [hunk 1]
```diff
@@ -604,3 +604,17 @@ spark.yarn.am.extraJavaOptions -Dsun.security.krb5.debug=true -Dsun.security.spn
 
 Finally, if the log level for `org.apache.spark.deploy.yarn.Client` is set to `DEBUG`, the log
 will include a list of all tokens obtained, and their expiry details
+
+## Using the Spark History Server to replace the Spark Web UI
+
+It is possible to use the Spark History Server application page as the tracking URL for running
+applications in scenarios where it may be desired to disable the built-in application UI. Two steps
```
Comment (squito) [inline docs/running-on-yarn.md]: nit: first sentence reads a little funny.  maybe rephrase to:

It is possible to use the Spark History Server application page as the tracking URL for running
 applications where built-in application UI is disabled.  This may be desirable on secure clusters or to avoid the memory usage on the driver from the UI.

Up to you.  maybe doens't even need the second sentence.

**Review by squito — APPROVED** [review]
lgtm
very minor comments

**Comment by vanzin** [comment]
Merging to master.

@tgravescs we can address any feedback you might have when you're back.
